src/log/lss.rs

```diff
@@ -10,18 +12,42 @@ use super::*;
 /// `LockFreeLog` is responsible for putting data on disk, and retrieving
 /// it later on.
 pub struct LockFreeLog {
-    pub(super) iobufs: IOBufs,
+    pub(super) iobufs: Arc<IOBufs>,
+    flusher_shutdown: Arc<AtomicBool>,
+    flusher_handle: Option<std::thread::JoinHandle<()>>,
 }
 
 unsafe impl Send for LockFreeLog {}
 unsafe impl Sync for LockFreeLog {}
 
+impl Drop for LockFreeLog {
+    fn drop(&mut self) {
+        self.flusher_shutdown.store(true, std::sync::atomic::Ordering::SeqCst);
+        if let Some(join_handle) = self.flusher_handle.take() {
+            join_handle.join().unwrap();
```

Comment:
we call `join()` here to wait until the flusher thread wakes up, reads the flusher_shutdown atomic bool (now set to true) and returns. we need to do this because the flusher thread still has a reference to the underlying `IOBufs`. This is important because in the impl of `Drop` for `IOBufs`, it calls `self.flush()` for each of its buffers, making sure all pending data is written to disk. If we return from this destructor without making sure that the flusher thread has finished (dropping the Arc count to 1, for the `LockFreeLog` which has the other reference) then the data may not be flushed by the time the process quits. This basically forces flushing of pending writes before the `LockFreeLog` itself is dropped.

This was causing a panic before I did this, in  `tests/test_pagecache.rs (basic_recovery test)`  because the old log was being dropped before data was flushed, then a new pagecache was created, and it tried to read a page that should have been present, but it hadn't been flushed yet, and its test failed.